src/test/perf/msgpass/msgpass.cc

```diff
@@ -0,0 +1,302 @@
+/**
+ * A simulation of a message-passing application workload for snmalloc.
+ *
+ * - N_PRODUCER producer threads allocate and queue spans of messages randomly,
+ * - to N_CONSUMER consumer threads, which dequeue messages and free() them.
+ *
+ * Optionally, N_PROXY threads act as both producers and consumers, forwarding
+ * received messages back to another queue rather than freeing them.
+ */
+
+#include "test/opt.h"
+#include "test/setup.h"
+#include "test/usage.h"
+#include "test/xoroshiro.h"
+
+constexpr static bool be_chatty = false;
+
+#include <chrono>
+#include <iomanip>
+#include <iostream>
+#include <snmalloc/snmalloc.h>
+#include <stdarg.h>
+#include <thread>
+#include <vector>
+
+using namespace snmalloc;
+
+void chatty(const char* p, ...)
+{
+  if constexpr (be_chatty)
+  {
+    va_list va;
+    va_start(va, p);
+    vfprintf(stderr, p, va);
+    va_end(va);
+  }
+}
+
+/*
+ * Interpret SNMALLOC_PASS_THROUGH ourselves to make this a bit more fair of a
+ * comparison, since relying of snmalloc itself to do the passing through
+ * results in it imposing its own idea of alignment onto the underlying
+ * allocator, which might result in it taking less optimized paths.
+ */
+#ifdef SNMALLOC_PASS_THROUGH
+struct MyAlloc
+{
+  MyAlloc() {}
+  void* alloc(size_t sz)
+  {
+    return malloc(sz);
+  }
+  void dealloc(void* p)
+  {
+    free(p);
+  }
+};
+#else
+struct MyAlloc
+{
+  snmalloc::Alloc& a;
+  MyAlloc() : a(ThreadAlloc::get()) {}
+  void* alloc(size_t sz)
+  {
+    return a.alloc(sz);
+  }
+  void dealloc(void* p)
+  {
+    a.dealloc(p);
+  }
+};
+#endif
```

Comment:
Could we just use malloc and free in the example and do something like

```C++
#ifndef SNMALLOC_PASS_THROUGH
#include <snmalloc/override/malloc.cc>
#endif
```

I have been wondering about doing this for more of the examples, so that the tests could be less dependent on snmalloc.

Reply:
Could do.  The ability to grab the thread allocator once per thread in the statically-`snmalloc` case seemed kind of nice (and you can get the dynamic behavior by using the malloc-passthrough and `LD_PRELOAD`-ing `snmallocshim`), but maybe it's not actually useful.